src/main/java/org/apache/datasketches/InequalitySearch.java

```diff
@@ -0,0 +1,200 @@
+ * search algorithm reaches the ends of the search range, the search algorithm calls the
+ * <i>resolve()</i> method to determine what to return to the caller.
+ * </li>
+ * <li><b>EQ:</b>  Find the adjacent pair <i>{A, B}</i> such that:<br>
+ * <i>arr[A] <= v <= arr[B]</i>. We return the index <i>A</i> or <i>B</i> whichever
+ * equals <i>v</i>, otherwise we return -1.
+ * </li>
+ * <li><b>GE:</b>  Find the lowest ranked adjacent pair <i>{A, B}</i> such that:<br>
+ * <i>arr[A] < v <= arr[B]</i>. Normally we return the index <i>B</i>. However if the
+ * search algorithm reaches the ends of the search range, the search algorithm calls the
+ * <i>resolve()</i> method to determine what to return to the caller.
+ * </li>
+ * <li><b>GT:</b>  Find the lowest ranked adjacent pair <i>{A, B}</i> such that:<br>
+ * <i>arr[A] <= v <= arr[B]</i>. Normally we return the index <i>B</i>. However if the
+ * search algorithm reaches the ends of the search range, the search algorithm calls the
+ * <i>resolve()</i> method to determine what to return to the caller.
+ * </li>
+ * </ul>
+ *
+ * @author Lee Rhodes
+ */
+public class InequalitySearch {
+
+  /**
+   * The enumerator of inequalities
+   */
+  public enum Inequality {
+
+    /**
+     * Less-Than
+     */
+    LT,
+
+    /**
+     * Less-Than Or Equals
+     */
+    LE,
+
+    /**
+     * Equals. Although not an inequality, it is included for completeness.
+     */
+    EQ,
+
+    /**
+     * Greater-Than Or Equals
+     */
+    GE,
+
+    /**
+     * Greater-Than
+     */
+    GT
+  }
+
+  /**
+   * Constructs this class
+   */
+  public InequalitySearch() { }
+
+  /**
+   * Binary Search for the index of the generic value in the given search range that satisfies
+   * the given inequality.
+   * If -1 is returned there are no values in the search range that satisfy the inequality.
+   *
+   * @param arr the given array that must be sorted.
+   * @param low the index of the lowest value in the search range
+   * @param high the index of the highest value in the search range
+   * @param v the value to search for.
+   * @param inequality one of LT, LE, EQ, GE, GT
+   * @param comparator for the type T
+   * @param <T> The generic type of value to be used in the search process.
+   * @return the index of the value in the given search range that satisfies the inequality.
+   */
+  public static <T> int find(final T[] arr, final int low, final int high, final T v,
+      final Inequality inequality, final Comparator<T> comparator) {
+    int lo = low;
+    int hi = high - 1;
+    int ret;
+    while (lo <= hi) {
+      final int midA = lo + (hi - lo) / 2;
+      ret = compare(arr, midA, midA + 1, v, inequality, comparator);
+      if (ret == -1 ) { hi = midA - 1; }
+      else if (ret == 1) { lo = midA + 1; }
+      else  { return getIndex(arr, midA, midA + 1, v, inequality, comparator); }
+    }
+    return resolve(lo, hi, low, high, inequality);
+  }
```

Comment:
Generics are always poorer in performance, but I developed this method as a preview of what will be a fully generic version of the ReqSketch, and how this can be implemented using standard Comparator<T> so that this could be more easily implemented in C++.